Add initial beatmap submission support

bdach opened the pull request:
https://github.com/user-attachments/assets/af71333d-1748-4a75-9b4d-393877b59acd

---

- Closes https://github.com/ppy/osu/issues/18901, I guess
- [x] Depends on https://github.com/ppy/osu/pull/31804

There will probably be follow-up fixes required for this, but this should be at least a minimal start required to start testing on staging.

Guest difficulty submission is not yet supported or exposed; this will require significant reworking of metadata management across the entire game first so that guest mappers become a first-class citizen everywhere.

bdach committed:
Add completion flag to `WizardOverlay`

bdach committed:
Implement special exporter intended specifically for submission flows

bdach committed:
Add beatmap submission support

Joehuu commented on osu.Game/Localisation/BeatmapSubmissionStrings.cs:
```diff
@@ -39,6 +39,31 @@ public static class BeatmapSubmissionStrings
+        /// </summary>
+        public static LocalisableString ConfirmSubmission => new TranslatableString(getKey(@"confirm_submission"), @"Submit beatmap!");
+
+        /// <summary>
+        /// "Exporting beatmap set in compatibility mode..."
+        /// </summary>
+        public static LocalisableString ExportingBeatmapSet => new TranslatableString(getKey(@"exporting_beatmap_set"), @"Exporting beatmap set in compatibility mode...");
+
+        /// <summary>
+        /// "Preparing beatmap set online..."
+        /// </summary>
+        public static LocalisableString PreparingBeatmapSet => new TranslatableString(getKey(@"preparing_beatmap_set"), @"Preparing beatmap set online...");
+
+        /// <summary>
+        /// "Uploading beatmap set contents..."
+        /// </summary>
+        public static LocalisableString UploadingBeatmapSetContents => new TranslatableString(getKey(@"uploading_beatmap_set_contents"), @"Uploading beatmap set contents...");
```
Any reason these strings have "set" and not the others? The term "beatmap set" only appears in code comments/logs and I believe just "beatmap" works fine here.

If intentional, going by the wiki, it should be "beatmapset" as stated "sometimes called" in https://osu.ppy.sh/wiki/en/Beatmap.

peppy replied:
https://github.com/ppy/osu/pull/31805/commits/753eae426d7c33978621025424b8dd43081a31fc

bdach committed:
Merge branch 'bss/api-setup' into bss/the-actual-submission

bdach committed:
Remove unnecessary suppressions

peppy committed:
Merge branch 'master' into bss/the-actual-submission

peppy committed:
Fix missing cached settings in `BetamapSubmissionOverlay` test

peppy committed:
Disallow changing beatmap / ruleset while submitting beatmap

peppy committed:
Don't show informational screens for subsequent submissions

These are historically only presented to the user when uploading a new
beatmap for the first time.

peppy committed:
Adjust beatmap stream creation to make a bit more sense

peppy committed:
Change `BeatmapSubmissionScreen` to use global back button instead of custom implementation

peppy committed:
Adjust timeouts to be much higher for upload requests

It seems that right now these timeouts do not check for actual data
movement, which is to say if a user with a very slow connection is
uploading and it takes more than `Timeout`, their upload will fail.

For now let's set these values high enough that most users will not be
affected.

peppy committed:
Update strings

peppy commented:
@bdach I've pushed an initial round of fixes. I'll do a second pass when the flow can be tested on staging, but seems pretty decent.

bdach committed:
Merge branch 'master' into bss/the-actual-submission

bdach committed:
Add staging submission service URL to development endpoint config

bdach committed:
Fix exiting being eternally blocked after successful beatmap submission

peppy commented:
We might want some better messaging when errors occur. These can come after the merge of this initial work.

A few common scenarios I can think of:

## Uploading a beatmap with incorrect online IDs:

![osu! 2025-02-10 at 05 59 20](https://github.com/user-attachments/assets/9ff1b432-9b55-46c2-89dd-ce4f60151d28)

This isn't handled great in stable, but maybe we can offer a path to repair (like giving the user an option to upload as a new set, or attempting to link to the real upload, if it's really theirs).

## Uploading someone else's beatmap

![osu! 2025-02-10 at 06 01 34](https://github.com/user-attachments/assets/b87abf06-6d1b-4f12-8f95-9a1ffe153a87)

Currently errors very late into the process, and with a nondescript error message.

peppy committed:
Add delay before browser displays beatmap

peppy committed:
Fix back button displaying before it should

peppy committed:
Adjust animations and induce some short delays to make things more graceful

peppy commented:
I get this won't be a huge improvement (it still looks horrible...), but is there a reason we are going all the way back to menu for the present operation?

ie can we do this instead?

```diff
diff --git a/osu.Game/Screens/Edit/Submission/BeatmapSubmissionScreen.cs b/osu.Game/Screens/Edit/Submission/BeatmapSubmissionScreen.cs
index 121e25d8b7..9672e4360a 100644
--- a/osu.Game/Screens/Edit/Submission/BeatmapSubmissionScreen.cs
+++ b/osu.Game/Screens/Edit/Submission/BeatmapSubmissionScreen.cs
@@ -28,7 +28,7 @@
 using osu.Game.Online.API.Requests.Responses;
 using osu.Game.Overlays;
 using osu.Game.Overlays.Notifications;
-using osu.Game.Screens.Menu;
+using osu.Game.Screens.Select;
 using osuTK;
 
 namespace osu.Game.Screens.Edit.Submission
@@ -418,7 +418,7 @@ public override bool OnExiting(ScreenExitEvent e)
                     }
 
                     s.Push(new EditorLoader());
-                }, [typeof(MainMenu)]);
+                }, [typeof(SongSelect)]);
 
                 return false;
             }

```

bdach committed:
Allow performing beatmap reload after submission from song select

bdach committed:
Remove unused using

peppy committed:
Minor code refactors

bdach committed:
Merge branch 'master' into bss/the-actual-submission
